samples/config_helper.cc

```diff
@@ -0,0 +1,487 @@
+
+  for (uint32_t i = 0; i < count; ++i) {
+    if (properties[i].queueFlags &
+        (VK_QUEUE_GRAPHICS_BIT | VK_QUEUE_COMPUTE_BIT)) {
+      return i;
+    }
+  }
+
+  return std::numeric_limits<uint32_t>::max();
+}
+
+}  // namespace
+
+ConfigHelper::ConfigHelper() = default;
+
+ConfigHelper::~ConfigHelper() = default;
+
+amber::Result ConfigHelper::CreateInstance() {
+  if (instance_ != VK_NULL_HANDLE)
+    return amber::Result("Sample: Vulkan instance already exists");
+
+  VkApplicationInfo app_info = {};
+  app_info.sType = VK_STRUCTURE_TYPE_APPLICATION_INFO;
+  app_info.apiVersion = VK_MAKE_VERSION(1, 0, 0);
+
+  VkInstanceCreateInfo instance_info = {};
+  instance_info.sType = VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO;
+  instance_info.pApplicationInfo = &app_info;
+
+  if (vkCreateInstance(&instance_info, nullptr, &instance_) != VK_SUCCESS)
+    return amber::Result("Sample: vkCreateInstance Fail");
+
+  return {};
+}
+
+amber::Result ConfigHelper::ChoosePhysicalDevice() {
+  uint32_t count;
+  std::vector<VkPhysicalDevice> physical_devices;
+
+  if (vkEnumeratePhysicalDevices(instance_, &count, nullptr) != VK_SUCCESS)
+    return amber::Result("Sample: vkEnumeratePhysicalDevices Fail");
+  physical_devices.resize(count);
+  if (vkEnumeratePhysicalDevices(instance_, &count, physical_devices.data()) !=
+      VK_SUCCESS) {
+    return amber::Result("Sample: vkEnumeratePhysicalDevices Fail");
+  }
+
+  for (uint32_t i = 0; i < count; ++i) {
+    if (!AreAllRequiredFeaturesSupported(physical_devices[i])) {
+      continue;
+    }
+
+    if (!AreAllExtensionsSupported(physical_devices[i])) {
+      continue;
+    }
+
+    queue_family_index_ = ChooseQueueFamilyIndex(physical_devices[i]);
+    if (queue_family_index_ != std::numeric_limits<uint32_t>::max()) {
+      physical_device_ = physical_devices[i];
+      return {};
+    }
+  }
+
+  return amber::Result("Vulkan::No physical device supports Vulkan");
+}
+
+amber::Result ConfigHelper::CreateDevice() {
+  if (device_ != VK_NULL_HANDLE)
+    return amber::Result("Sample: Vulkan device already exists");
+
+  VkDeviceQueueCreateInfo queue_info;
+  const float priorities[] = {1.0f};
+
+  queue_info.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
+  queue_info.queueFamilyIndex = queue_family_index_;
+  queue_info.queueCount = 1;
+  queue_info.pQueuePriorities = priorities;
+
+  VkDeviceCreateInfo info = {};
+  info.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
+  info.pQueueCreateInfos = &queue_info;
+  info.queueCreateInfoCount = 1;
+  info.pEnabledFeatures = &kRequiredFeatures;
+  info.enabledExtensionCount =
+      static_cast<uint32_t>(kNumberOfRequiredExtensions);
+  info.ppEnabledExtensionNames = kRequiredExtensions;
+
+  if (vkCreateDevice(physical_device_, &info, nullptr, &device_) != VK_SUCCESS)
+    return amber::Result("Vulkan::Calling vkCreateDevice Fail");
+
+  return {};
+}
+
+std::pair<amber::Result, amber::VulkanEngineConfig>
+ConfigHelper::CreateVulkanConfig() {
```

Comment:
Why not amber::Result COnfigHelper::CreateVulkanConfig(amber::VulkanEngineConfig* config) then we just fill in the config?